src/lib/transition_router/transition_router.ml

```diff
@@ -168,7 +168,58 @@ let start_bootstrap_controller ~context:(module Context : CONTEXT) ~trust_system
     match Envelope.Incoming.sender block with Remote r -> [ r ] | Local -> []
   in
   let preferred_peers = Option.value_map ~f ~default:[] best_seen_transition in
+  (* TODO find a better place for this *)
+  let fetch_completed_snarks () =
+    Deferred.List.iter
+      ~f:(fun peer ->
+        let completed_works =
+          Mina_networking.get_completed_checked_snarks network peer
+        in
+        let%bind completed_works = completed_works in
+        let completed_works =
+          match completed_works with
+          | Error e ->
+              [%log error]
+                ~metadata:
+                  [ ("peer", Network_peer.Peer.to_yojson peer)
+                  ; ("error", Error_json.error_to_yojson e)
+                  ]
+                "Failed to fetch completed snarks from peer: $error" ;
+              []
+          | Ok completed_works ->
+              completed_works
+        in
+        let () =
+          List.iter completed_works ~f:(fun work ->
+              let callback =
+                let cb =
+                  Mina_net2.Validation_callback.create_without_expiration ()
+                in
+                Network_pool.Snark_pool.Broadcast_callback.External cb
+              in
+              (* proofs should be verified in apply and broadcast *)
+              let msg =
+                let statement = Transaction_snark_work.Checked.statement work in
+                let snark =
+                  Network_pool.Priced_proof.
+                    { proof = work.proofs
+                    ; fee = { fee = work.fee; prover = work.prover }
+                    }
+                in
```

Comment:
I don't fully understand the callback logic, so there is no way i did this right.

Reply:
Hmm, this doesn't look right to me either. You are rebroadcasting all of the work. You should be able to call the apply part separately. You also need to verify the proofs-
I think the following steps need to be done:
1. Get all the completed work from peers
2. Check if there any lower fee snarks already in the pool in which case skip
3. Check if the work sent is actually referenced in the frontier (and that they are not some random snarks statements)
4. verify the proofs (batch verification). Steps 2, 3, 4 is mostly what's in `Snark_pool.verify_and_act`
5. add it to the pool if verification is successful (`Snark_pool.add_snark`)

Seems like such a function should live in `Snark_pool`

Reply:
That makes sense; i can do that. The only reason i used this function was I wanted to keep expected behavior with received snark work, which is currently broadcasting to the gossip net when something completed was received. The apply and broadcast function also had the added bonus of verifying so I figured it was the natural move.

Reply:
> The apply and broadcast function also had the added bonus of verifying so I figured it was the natural move.

afaiu verification is a separate step. Look up `verify_impl` in `pool_sink.ml`

Reply:
done